### ISSUE TITLE
Controller Config resets everytime at start

### ISSUE THREAD

**Issue by salteax** [m64p/m64p#248]
I'm using a original N64 Controller connected via USB with a raphnet cable, my problem is that the config of the emulator auto sets to wrong settings even when i try to configure the settings the right way it always overrides them. How can i fix this problem?

**Comment by loganmc10** [comment]
Were you able to figure this out? After you make a controller profile, you need to assign it to your controller, or it will continue to us the "Auto" profile

**Comment by salteax** [comment]
No sadly not, i might try again and give a precise explanation

**Comment by JarleSoelie** [comment]
Hi, I'm having a similar problem I think. The Config "input-settings.ini" keeps resetting and setting the Gamepad=Auto and preventing my actual controller from working. However, this only seems to happen when I start mupen64plus through LaunchBox (a emulation frontend). 
I tried to set the input-settings.ini to "Read only", but still mupen64plus manages to revert the config to Gamepad=Auto when I run the emulator through LaunchBox. 
Maybe worth noting, I am also using the latest DS4Windows with a profile that adds some keyboard keys (F, F5, and F7) to controller buttons when I start the emulator. I have no trouble with other emulators or games via LaunchBox, and everything was also functioning just fine before I updated mupen64plus-gui. Sadly I don't know what build I was using before I updated, but it was from earlier this year. I will try and do some more testing without DS4Windows. I really hope to find a solution to this. Thanks for your time, I really like this emulator!

**Comment by loganmc10** [comment]
@JarleSoelie thanks for the excellent description. It sounds like this is happening because it can't contact the controller. The way the input plugin is written, if you've selected a specific controller, and it can't find it, it reverts to Auto, hoping it can find a controller for you that will work.

I'm guessing since you say it only happens when using LaunchBox, that maybe LaunchBox has a lock/hold on the controller that isn't released in time for m64p to pick it up?

**Comment by JarleSoelie** [comment]
Thanks for the quick response, I think I figured out what was wrong and it's mostly a me problem... 
I am using DS4Windows with a DualShock 4 controller emulated as a DualShock 4 controller (not as a Xbox controller). It seems the input plugin is not compatible with DualShock 4 controllers? At least no controller is recognized by mupen64plus if the controller is emulated as a Dualshock4 via DS4Windows. 

My preferred setup is that when I start mupen64plus, DS4Windows will autostart a profile with my previously mentioned hotkeys and also turn into a Xbox controller (that is recognized by mupen64plus). This seems to function just fine when running the emulator as a standalone program, but when I start it via LaunchBox, mupen64plus loses sight of the controller and reverts to auto. However, this doesn't happen every time!.. Sometimes it will work just fine, but then some times between switching games it will just revert to auto. It is my impression that the emulator will start before DS4Windows has a chance to change the profile and thus mupen will think there is no controller connected and revert to auto, which in turn does not detect my controller via DS4Windows... So I think you are absolutely unto something when saying that the controller isn't released in time.

Long story short, emulated DS4 controllers via DS4Windows are not detectable by mupen64plus? I guess there is no easy way to modify the input plugin with an option to prevent something like this? Like an option to keep a manual controller in place with no auto override, or make the plugin compatible with DS4Windows and DS4 controllers? Either way, this is likely a specialized "me" problem and not worth spending that much time on, or risk cluttering up the GUI with options. 

Thanks again! And sorry if all this was out of place here

**Comment by loganmc10** [comment]
The input is handled by SDL2, so whatever controllers are supported by SDL2 should be supported by m64p, I don't really have any control over that part.

> Like an option to keep a manual controller in place with no auto override

This is a little tricky, since the list of available controllers is populated via SDL2, so if SDL2 doesn't recognize the controller, there is nothing to "keep" there. I know the list is populated by names, but actually what the input plugin is remembering is an index number, which SDL2 resolves into a name. So either way it really depends on the controller being present for that list to be populated correctly.

**Comment by JarleSoelie** [comment]
Thank you for the insight! I'll live with this for now. 
I imagine a solution to issue 207 would also fix my problem, since DS4Windows does connect a controller that the plugin recognizes, it just seems to do this too late and thus after mupen64plus has started. But still, I guess this is easier said than done as well.
https://github.com/m64p/m64p/issues/207#issue-958445512
Thanks again!

**Comment by Jj0YzL5nvJ** [comment]
From what I understand, the solution is to add a delay (`ping -n 3 127.0.0.1 >nul`) of 2 or 3 seconds before launching m64p from the LaunchBox.

**Comment by JarleSoelie** [comment]
> From what I understand, the solution is to add a delay (`ping -n 3 127.0.0.1 >nul`) of 2 or 3 seconds before launching m64p from the LaunchBox.

Thanks for the suggestion! I thought about that too, but DS4Windows doesn't switch to the profile with the hotkeys (and Xbox controller) before it sees that the m64p process has started, and by then it tends to be too late and m64p is back to auto. A testament to the fast startup of m64p I guess. Unless your suggested ping delay appears after the m64p process first starts? 

I could create a batch file via launchbox with commands that change the DS4Windows profile to one with Xbox and then start m64p afterwards, but this seems to cause launchbox to not count the game time (because as soon as the bat files is done, Launchbox thinks the game is done)... Life is so hard... 

Anyway! My current workaround, if anyone is curious, is to not use the DS4 emulated controller before starting m64p at all. I let DS4Windows by default emulate the controller as a xbox controller and not as a DS4. Which in practice meant that I had to flip all my auto profiles in DS4Windows from DS4 to Xbox and create new auto-profiles for games that work better with DS4. In other words no more default DS4 controller for my setup

**Comment by DarthSidious666** [comment]
With the current build you don't need to use any 3rd party apps to do a dualshock 4 controller.
I've tested this recently and as long as you have a profile set up for it it will work without issues.
